pisa/population.py

```diff
@@ -0,0 +1,206 @@
+import urllib
+from dataclasses import dataclass
+from enum import Enum
+
+import numpy as np
+import pandas as pd
+import requests
+import rasterio
+from rasterio.mask import mask
+from geopandas import GeoDataFrame, clip, points_from_xy
+from shapely import MultiPolygon, Polygon
+
+from hdx.api.configuration import Configuration
+from hdx.data.resource import Resource
+
+@dataclass
+class Population:
+
+    data_source: str
+    iso3_country_code: str
+    admin_area_boundaries: Polygon | MultiPolygon
+    population_resolution: int = 5
+
+    class DataSource(Enum):
+        WORLD_POP = 'world_pop'
+        FACEBOOK = 'facebook'
+
+    def get_population_gdf(self) -> GeoDataFrame:
+        """Integrates the methods to get the population numbers for the selected area into one flow and
+        returns grouped population data for the admin area as a GeoDataFrame."""
+
+        population_df = self.get_population_data_with_selected_method()
+
+        return self.group_population(population_df, self.population_resolution)
+
+    def get_population_data_with_selected_method(self) -> pd.DataFrame:
+        """Select correct method according to the data_source given and return the population data as a DataFrame."""
+        if self.data_source == self.DataSource.WORLD_POP.value:
+            return self.get_population_worldpop()
+        elif self.data_source == self.DataSource.FACEBOOK.value:
+            return self.get_population_facebook()
+        else:
+            raise ValueError(
+                f"Invalid data source: {self.data_source}. Must be one of {[ds.value for ds in Population.DataSource]}.")
+
+    @staticmethod
+    def group_population(population_df: pd.DataFrame, population_resolution: int) -> GeoDataFrame:
+        """ Group population data by longitude and latitude and return as a GeoDataFrame."""
```

Comment:
Might be worth explaining what the population resolution is and what does it mean to group by lat and lon. I think this question was raised in the past, so I can assume that in the future someone else will wonder :p